src/main/kotlin/org/rust/cargo/runconfig/target/RsCommandLineSetup.kt

```diff
@@ -0,0 +1,117 @@
+import com.intellij.execution.target.value.TargetValue
+import com.intellij.execution.target.value.getUploadRootForLocalPath
+import com.intellij.lang.LangBundle
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.util.io.isDirectory
+import org.jetbrains.concurrency.AsyncPromise
+import org.jetbrains.concurrency.Promise
+import org.rust.stdext.toPath
+import java.nio.file.Path
+import java.nio.file.Paths
+
+@Suppress("UnstableApiUsage")
+class RsCommandLineSetup(val request: TargetEnvironmentRequest) {
+    private val languageRuntime: RsLanguageRuntimeConfiguration? = request.configuration?.languageRuntime
+    private val environmentPromise = AsyncPromise<Pair<TargetEnvironment, TargetProgressIndicator>>()
+    private val dependingOnEnvironmentPromise: MutableList<Promise<Unit>> = mutableListOf()
+    private val uploads: MutableList<Upload> = mutableListOf()
+    private val projectHomeOnTarget: VolumeDescriptor = VolumeDescriptor(
+        RsCommandLineSetup::class.java.simpleName + ":projectHomeOnTarget",
+        "",
+        "",
+        "",
+        request.projectPathOnTarget
+    )
+
+    fun requestUploadIntoTarget(uploadPathString: String): TargetValue<String> {
+        val uploadPath = FileUtil.toSystemDependentName(uploadPathString).toPath()
+        val isDir = uploadPath.isDirectory()
+        val localRootPath = if (isDir) uploadPath else (uploadPath.parent ?: Paths.get("."))
+        val (uploadRoot, pathToRoot) = request.getUploadRootForLocalPath(localRootPath.toString())
+            ?: createUploadRoot(projectHomeOnTarget, localRootPath).let { uploadRoot ->
+                request.uploadVolumes += uploadRoot
+                uploadRoot to "."
+            }
+        val result = DeferredTargetValue(uploadPathString)
+        dependingOnEnvironmentPromise += environmentPromise.then { (environment, targetProgressIndicator) ->
+            if (targetProgressIndicator.isCanceled || targetProgressIndicator.isStopped) {
+                result.stopProceeding()
+                return@then
+            }
+            val volume = environment.uploadVolumes.getValue(uploadRoot)
+            try {
+                val relativePath = if (isDir) pathToRoot
+                else uploadPath.fileName.toString().let {
+                    if (pathToRoot == ".") it else joinPath(pathToRoot, it)
+                }
+                val resolvedTargetPath = volume.resolveTargetPath(relativePath)
+                uploads.add(Upload(volume, relativePath))
+                result.resolve(resolvedTargetPath)
+            } catch (t: Throwable) {
+                LOG.warn(t)
+                targetProgressIndicator.stopWithErrorMessage(
+                    LangBundle.message(
+                        "progress.message.failed.to.resolve.0.1",
```

Comment:
`'progress.message.failed.to.resolve.0.1' doesn't appear to be a valid property key` warning, maybe `LangCoreBundle` should be used instead of `LangBundle`